grub-core/loader/i386/txt/acmod.c

```diff
@@ -0,0 +1,610 @@
+    return NULL;
+
+  /* Not checking if (hdr + elem_end) overflows. We know that (hdr + acm_len)
+   * doesn't, and that elem_end <= acm_len. For the same reason we don't have to
+   * check if (hdr + offset) overflows. */
+
+  return (void *)((unsigned long)hdr + offset);
+}
+
+static void*
+fits_in_acm (struct grub_txt_acm_header *hdr, grub_uint32_t offset,
+	     grub_uint32_t size)
+{
+  return n_fit_in_acm(hdr, offset, size, 1);
+}
+
+/*
+ * Returns pointer to ACM information table. If the table is located outside of
+ * ACM or its reported size is too small to cover at least 'length' field,
+ * NULL is returned instead.
+ */
+static struct grub_txt_acm_info_table*
+get_acmod_info_table (struct grub_txt_acm_header* hdr)
+{
+  grub_uint32_t user_area_off, info_table_size;
+  struct grub_txt_acm_info_table *ptr = NULL;
+  /* Minimum size required to read full size of table */
+  info_table_size = __builtin_offsetof (struct grub_txt_acm_info_table, length)
+		    + sizeof(ptr->length);
+
+  /* Overflow? */
+  if ( grub_add (hdr->header_len, hdr->scratch_size, &user_area_off) )
+    {
+      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM header length plus scratch size overflows"));
+      return NULL;
+    }
+
+  if ( grub_mul (user_area_off, 4, &user_area_off) )
+    {
+      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM header length and scratch size in bytes overflows"));
+      return NULL;
+    }
+
+  ptr = fits_in_acm(hdr, user_area_off, info_table_size);
+
+  if ( ptr != NULL )
+    {
+      if ( info_table_get (ptr, length) < info_table_size )
+	return NULL;
+
+      info_table_size = info_table_get (ptr, length);
+      ptr = fits_in_acm(hdr, user_area_off, info_table_size);
+    }
+
+  return ptr;
+}
+
+/*
+ * Function returns pointer to chipset ID list, after checking that
+ * grub_txt_acm_chipset_id_list and all grub_txt_acm_chipset_id structures are
+ * within ACM. Otherwise, NULL is returned.
+ */
+static struct grub_txt_acm_chipset_id_list*
+get_acmod_chipset_list (struct grub_txt_acm_header *hdr)
+{
+  struct grub_txt_acm_info_table *info_table;
+  grub_uint32_t id_entries_off;
+  struct grub_txt_acm_chipset_id_list *chipset_id_list;
+
+  /* This fn assumes that the ACM has already passed the is_acmod() checks */
+
+  info_table = get_acmod_info_table (hdr);
+  if ( info_table == NULL )
+    {
+      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM info table out of bounds"));
+      return NULL;
+    }
+
+  chipset_id_list = fits_in_acm(hdr, info_table_get (info_table, chipset_id_list),
+				sizeof(struct grub_txt_acm_chipset_id_list));
+  if ( chipset_id_list == NULL )
+    {
+      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM chipset ID list out of bounds"));
+      return NULL;
+    }
+
+  /* Overflows? */
+  if ( grub_add (info_table->chipset_id_list,
+		 sizeof(struct grub_txt_acm_chipset_id_list), &id_entries_off))
+    {
+      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM chipset ID entries out of bounds"));
+      return NULL;
+    }
```

Comment:
`fits_in_acm()` checked this already.

```suggestion
  id_entries_off = info_table->chipset_id_list + sizeof(struct grub_txt_acm_chipset_id_list);
```